src/NosCore.Controllers/NpcPacketController.cs

```diff
@@ -135,10 +155,144 @@ public void Shopping(ShoppingPacket shoppingPacket)
                 return;
             }
 
-            var shopRate = Session.Character.GenerateShopRates();
+         
             Session.SendPacket(aliveEntity.GenerateNInv(shopRate.Item1, shoppingPacket.ShopType, shopRate.Item2));
         }
 
+        public void CreateShop(MShopPacket mShopPacket)
+        {
+            if (Session.Character.InExchangeOrTrade)
+            {
+                //todo log
+                return;
+            }
+
+            if (Session.Character.MapInstance.Portals.Any(por
+                => Session.Character.PositionX < por.SourceX + 6
```

Comment:
Making a method to check in a circle could be useful for later (quests for example)

Reply:
i don't think portal deserve it. it's not a visual entity